### ISSUE TITLE
Element-R | After reseting backup on big accounts, Element will stop syncing, sending message until all keys are marked as not backed-up

### ISSUE THREAD

**Issue by BillCarsonFr** [element-hq/element-web#26892]
## Step to reproduce

- Have an account with a lot of keys in local database, wait for all of them to be uploaded to backup
- Go to Security & Privacy and click reset, and follow the step

=> Starting now if you type a message, it won't be sent after several minutes depending on the number of room keys in database

## What is happening

See here https://github.com/element-hq/element-web/issues/26783#issuecomment-1867413139

Basically reseting backup will have to do a big write transaction that will lock the `inbound_group_sessions2` object store for several minutes.
It will also lock all the database due to the rust `save_changes_lock` that is taken when saving changes after a sync

=> Sync will be blocked, sending will be blocked

## Difference with legacy

The `inbound_group_sessions2` rust object store is bigger (encrypted byte array of lenght ~3700) than the legacy one (json with pickled key of around (string of lenght ~700).
So the rust db gets slower faster

**Comment by BillCarsonFr** [comment]
> Is this not a duplicate of #26873?

Could be one of the cause. This one is specifically after a backup reset.

**Comment by richvdh** [comment]
I actually meant https://github.com/element-hq/element-web/issues/26783.

But the issue is "Key sending is slow when there a lot of keys to back up", right? We should combine the two issues, imho.

**Comment by richvdh** [comment]
> But the issue is "Message sending is slow when there a lot of keys to back up", right? We should combine the two issues, imho.

no, I misunderstood. The issue here is that the actual transaction to reset the key backup takes a very long time. (And if you restart during that time, everything is messed up)

**Comment by andybalaam** [comment]
Plan:

* Change "needs_backup" field into "most_recent_backup_to_include_this" so we don't need to modify all records when the current backup version changes.
* Reduce the size of the values in inbound_group_sessions2 so everything is faster. Do this by avoiding so many JSON re-encodes.

**Comment by andybalaam** [comment]
Previous plan doesn't work without a migration because we can't efficiently query for records that **don't** have a matching `most_recent...` field.

Next idea: assign each backup version an order number, so each time we see a new backup version from the server, we set a `current_backup_order` value to the previous order + 1. We store against each inbound_group_session the order of the most recent backup that contains this session under e.g. `backup_order_including_this`. (0 == never backed up)

In the backup loop we query inbound_group_sessions where `backup_order_including_this < current_backup_order`.

The above would require a migration.

Since we don't want to combine this with a migration to reduce value size, we can avoid a further migration by preparing now.

Prep means: add another field to inbound_group_sessions called `legacy: true`, and set `backup_order_including_this: 0`. This means these records will appear in the backup loop, and we will migrate it during that loop (instead of in a migration).

Migrating during the backup loop means: check for `legacy: true`. If set, removed legacy:true, set `backup_order_including_this:1`. (When we start counting orders, we will assign the current one to be 1.) Then if the true current order is >1, we back up this session, otherwise we are done.

**Comment by BillCarsonFr** [comment]
Some remarks on 
> Next idea: assign each backup version an order number, so each time we see a new backup version from the server, we set a current_backup_order value to the previous order + 1. 

What happens in situations were new backups version are created (without deleting the previous ones) without the client having beeing on to be aware of them. Then we start to delete them.

Something like
- Create a version "1"
- Create a version "2"
- Create a version "3"
- Create a version "4"

The client is aware of version "1" and give it order 0
Then it misses version "2" and "3"
But get version "4", it will have order 1 then.

Now version "4" is deleted, making the old "3" as current.
The client gets to know it and will give it order 2

I guess it still works? but can there be some edge cases due to that?

**Comment by richvdh** [comment]
> Now version "4" is deleted, making the old "3" as current.
> The client gets to know it and will give it order 2

That's fine. The most recent backup is then version "3", and we will try to upload all of our keys to it. 

Of course, version "3" might already have all of our keys, so that would be somewhat wasted effort. But it'll sort itself out in the end.

**Comment by andybalaam** [comment]
Not blocking Element R release because if we do https://github.com/element-hq/element-web/issues/26930 we will be on-par with legacy crypto. (But we should still do this later.)

**Comment by BillCarsonFr** [comment]
FTR, with the new db format it took 3mn to reset all my keys (a bit less than 300k keys) instead of several 10s of minutes previously.

**Comment by andybalaam** [comment]
To speed up key backup resets:

* [ ] Store current backup order number and ID in Indexed DB, probably in the `core` store, or in `backup_keys`. Default it to 1 and the current backup ID if it is empty.
* [ ] When we receive a new backup version (or when we reset backup?), increment the current backup order number and change the current ID to match the new ID
* [ ] In the backup loop, follow the algorithm from my above comment

To remove legacy code and data:

* [ ] Wait until most users have probably migrated most of their keys
* [ ] Change the logic from the above comment to remove the first `if`. This means we don't need the `not_backed_up` property any more. Any non-migrated keys will be backed up again, but this won't do harm except being slow.
* [ ] Create a DB migration that removes the `not_backed_up` property and its related index.

**Comment by andybalaam** [comment]
I just had a conversation with @poljar and @richvdh about how we can implement this. Here are my notes from our conversation:

* Modify `inbound_group_sessions_for_backup` and `mark_inbound_group_sessions_as_backed_up` so they take in the backup version. We checked where they are called, and we do have access to the version in both places. It's stored somewhere deep inside `PendingBackupRequest`. We discussed and agreed that this is a sensible API change for all crypto storage backends, because in both cases we are talking about the "current backup", which was implicit, and this allows us to be more explicit (maybe even avoiding races) by saying exactly which backup we mean.
* Create a new object store `backup_versions` in Indexed DB that will hold a mapping of backup version -> backup order number. We decided we should keep all of them, not just the latest, so that if some old backup is still running we won't somehow understand that as a new backup and bump our version again
* In `mark_inbound_group_sessions_as_backed_up`:
    * check the version we have been given against the versions we know in `backup_versions`.
    * If it doesn't exist in there, add it, giving it an order number greater than all the other entries.
    * Look up the order number for this version, and write that into the `backed_up_to` field for each entry we are saving.
* In `inbound_group_sessions_for_backup`:
    * look up the version we have been given in `backup_versions`.
    * If it exists, use its order as `current_backup_order` in the algorithm from the above comment.
    * If it doesn't exist then this is a new backup: just return a batch of any entries, since none of them are backed up yet in this new backup. (The entry will get added when the batch succeeds, and `mark_inbound_group_sessions_as_backed_up` is called.)

Thinking through the case when this has never happened before:

* First we run a schema upgrade to add the `backup_versions` object store
* Then we call `inbound_group_sessions_for_backup` and find no entry for the current backup, so we would consider it new. We should have a special case: if `backup_versions` is empty, we should revert to the old algorithm, looking for `not_backed_up` entries.
* Then we call `mark_inbound_group_sessions_as_backed_up`. We will add an entry to `backup_versions` for this backup version. We should choose order number 1 because there are no other entries in the store. From now on things should work as planned because order number 1 means "the backup was current when we started this process" when we don't know better.

**Comment by BillCarsonFr** [comment]
Moving back to qualification following https://github.com/element-hq/element-web/issues/29192
